eclair-core/src/main/scala/fr/acinq/eclair/router/Graph.scala

```diff
@@ -98,49 +98,50 @@ object Graph {
+    val shortestPaths = new mutable.Queue[PathWithSpur]
+    shortestPaths.enqueue(PathWithSpur(WeightedPath(shortestPath, pathWeight(sourceNode, shortestPath, amount, currentBlockHeight, wr)), 0))
     // stores the candidates for the k-th shortest path, sorted by path cost
-    val candidates = new mutable.PriorityQueue[WeightedPath]
+    val candidates = new mutable.PriorityQueue[PathWithSpur]
 
     // main loop
     for (k <- 1 until pathsToFind) {
       if (!allSpurPathsFound) {
-        val prevShortestPath = shortestPaths(k - 1).path
-        // for every edge in the path, we will try to find a different path after that edge
-        for (i <- prevShortestPath.indices) {
-          // select the spur node as the i-th element of the previous shortest path
-          val spurNode = prevShortestPath(i).desc.a
-          // select the sub-path from the source to the spur node
-          val rootPathEdges = prevShortestPath.take(i)
+        val prevShortestPath = shortestPaths(k - 1).p.path
+        // for every new edge in the path, we will try to find a different path before that edge
+        for (i <- shortestPaths(k - 1).spurIndex until prevShortestPath.length) {
+          // select the spur node as the i-th element from the target of the previous shortest path
+          val spurNode = prevShortestPath(prevShortestPath.length - 1 - i).desc.b
+          // select the sub-path from the spur node to the target
+          val rootPathEdges = prevShortestPath.takeRight(i)
           // we ignore all the paths that we have already fully explored in previous iterations
-          // if for example the spur node is B, and we already found shortest paths starting with A-B-C and A-B-D,
-          // we want to ignore the B-C and B-D edges
-          //          +-- C -- [...]
-          //          |
-          // A -- B --+-- D -- [...]
-          //          |
-          //          +-- E -- [...]
-          val alreadyExploredEdges = shortestPaths.collect { case p if p.path.take(i) == rootPathEdges => p.path(i).desc }.toSet
-          // we also want to ignore any link that can lead back to the previous node (we only want to go forward)
-          val returningEdges = rootPathEdges.lastOption.map(last => graph.getEdgesBetween(last.desc.b, last.desc.a).map(_.desc).toSet).getOrElse(Set.empty)
+          // if for example the spur node is D, and we already found shortest paths ending with A->D->E and B->D->E,
+          // we want to ignore the A->D and B->D edges
+          // [...] --> A --+
+          //               |
+          // [...] --> B --+--> D --> E
+          //               |
+          // [...] --> C --+
+          val alreadyExploredEdges = shortestPaths.collect { case p if p.p.path.takeRight(i) == rootPathEdges => p.p.path(p.p.path.length - 1 - i).desc }.toSet
+          // we also want to ignore any vertex on the root path to prevent loops
+          val alreadyExploredVertices = rootPathEdges.map(_.desc.b).toSet
```

Comment:
Isn't this too aggressive? Aren't we going to miss edges from a previously explored vertex that lead to a new alternative route?

Reply:
No, any spur path that passes through a node on the root path would create a loop since the same node is visited once in the spur path and then again in the root path. That's why it used to produce paths with loops.